File: crack-data-structures-and-algorithms/leetcode/coin_change_q322.py
# 核心思路
# 使用DP，状态转移方程
# dp[i] 表示金额为i时最少需要的零钱数
# c1, c2, c3 ... cn 代表零钱的面额
# dp[i] = min(dp[i-c1] + 1, dp[i-c2] + 1, ... dp[i-cn] + 1)
# dp[0] = 0
# 注意默认值 amount+1 同时也作为哨兵
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [amount+1] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            for c in coins:
                delta = i - c
                if delta < 0:
                    continue
                # Update dp[i] to minimum if possible.
                dp[i] = min(dp[i], dp[delta] + 1)

        return dp[-1] if dp[-1] != (amount + 1) else -1

File: crack-data-structures-and-algorithms/leetcode/coin_change_q322.cpp
#include <algorithm>
#include <cstdint>
#include <vector>

using namespace std;

// 核心思路
// 对于给定的 coins 列表，我们有
// dp[n] = min(dp[n-coins[i]] + 1, dp[n]), 0 <= i < |coins|
// dp[0] = 0
// 因为反复用到了 dp[n] 的的值，所以需要仔细考虑 dp 数组的初始值/哨兵值
// 这个值可以用 amount + 1 或者 -1 来承担
// 如果使用-1，那么min()比较时注意一定要转换成 unsigned 来比较，否则最小值一直都会是-1
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, -1);
        dp[0] = 0;
        for (int i = 1; i <= amount; ++i) {
            for (auto c : coins) {
                if (c <= i && dp[i-c] != -1) {
                    dp[i] = min<uint32_t>(dp[i], dp[i-c] + 1);
                }
            }
        }

        return dp[amount];
    }
};
